### ISSUE TITLE
Wrong colours in glb when DiffuseColor is set

### ISSUE THREAD

**Issue by Amoki** [IfcOpenShell/IfcOpenShell#2509]
I have a wall with light gray colour
 ```
IFCCOLOURRGB($,0.847058823529,0.8,0.745098039216);
IFCSURFACESTYLERENDERING(#97,0.,IFCNORMALISEDRATIOMEASURE(0.45),$,$,$,IFCNORMALISEDRATIOMEASURE(0.6),$,.NOTDEFINED.)
```
In the output gltf, the colour is a dark gray
```json
"pbrMetallicRoughness": {
    "baseColorFactor": [
        0.38117647058805,
        0.36000000000000004,
        0.3352941176472,
        1.0
    ],
    "metallicFactor": 0
}
```
I guess the DiffuseColor is applied to the colour: `0.847058823529 * 0.45 (the DiffuseColor) == 0.38117647058805`.

Is this an intended behavior? 
If I understand well, the DiffuseColor should be used by the 3D renderer engine, not pre-computed by `IfcConvert`.

If I compare the result with other IFC viewers, the wall is a light gray, not the dark gray outputted by `IfcConvert`

[Wall_without_DiffuseColor.ifc.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/9811499/Wall_without_DiffuseColor.ifc.txt)
[Wall_with_DiffuseColor.ifc.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/9811500/Wall_with_DiffuseColor.ifc.txt)

**Comment by Moult** [comment]
IFC (like other 3D software) differentiates between the surface colour used for basic viz (i.e. in a viewport) compared with the diffuse colour used for rendering (i.e. fed into a rendering engine to be calculated).

glTF does not differentiate - it is always the colour used for rendering, and in this example the reflectance method in glTF is physical (i.e. PBR). In IFC4, there was no enum available for specifying PBR, but I'd say it's a good fallback in this day and age when the reflectance method is set to "NOTDEFINED" (as it is in your exmaple).

So I'd say that IfcConvert is doing the correct thing in bringing the diffuse colour into PBR for glTF.

Other IFC viewers are just visualising the IFC, and so they show the wall as light grey which is the surface colour. This is also correct.

In the BlenderBIM Add-on, you actually can see both: in the viewport you see it as light grey, but the X3D/glTF compatible diffuse colour is stored too, so when you switch to rendered view, you see the dark grey.

**Comment by aothms** [comment]
Indeed, this happens here https://github.com/IfcOpenShell/IfcOpenShell/blob/4cccfd08d3f314d58dcdf5fde62dfc8a053d3fa1/src/ifcgeom/IfcGeom.cpp#L4092-L4095

In the 2X3/4 docs we had this:

>  The diffuse part of the reflectance equation can be given as either a colour or a scalar factor.
The diffuse colour field reflects all light sources depending on the angle of the surface with respect to the light source. The more directly the surface faces the light, the more diffuse light reflects.
The diffuse factor field specifies how much diffuse light from light sources this surface shall reflect. Diffuse light depends on the angle of the surface with respect to the light source. The more directly the surface faces the light, the more diffuse light reflects. **The diffuse colour is then defined by surface colour * diffuse factor.**

Which is a bit of a weird statement tbh, but I guess it's what I based myself on.

I also think multiplication is a reasonable approach.

But I guess we can introduce a flag like `--surface-colour` `--diffuse-colour` `--multiply-surface-diffuse` to choose the behaviour at runtime.

**Comment by Amoki** [comment]
Thanks for your answers!

All three viewers I tested (Solibri Anywhere, usBIM, BIMCollab Zoom) don't apply the diffuse-color.

> Diffuse light depends on the angle of the surface with respect to the light source. The more directly the surface faces the light, the more diffuse light reflects

As there is no light direction in IfcConvert, the pre-computed value always feels wrong.

A flag could be a solution!
Unfortunately, we have no skills in c++ and no dev env to try to submit a PR :/

**Comment by Moult** [comment]
All three viewers correctly don't apply the diffuse colour because all three viewers are not rendering engines, just "viewport" viewers.

glTF is however intended for rendering, and so applying the diffuse colour in glTF is the correct approach. The descriptions given in the IFC4 docs are pretty misleading and I did a bit of work upgrading it for IFC4.3. However, the intention in IFC4 and IFC2X3 is that it follows the X3D standard (and that sentence is copy pasted from X3D from memory). In other words, the diffuse colour is not a pre-computed value, it is truly a base colour, and so whether or not there is a light source is irrelevant. The key attribute that determines how the IFC DiffuseColour is rendered depends on the "ReflectanceMethod" attribute.

I'd support adding a flag if only to allow the glTF to be used as a way to transfer data into a non-rendered environment (or also to prevent confusion for those comparing against non-glTF viewport viewers), but the current behaviour sounds like a technically correct default.
